config/config.go

```diff
@@ -88,46 +104,127 @@ var defaultConfig = &config{
+		currentConfig, currentConfigErrors = Load(nil)
+		// return cfg, err
+	}
+	return currentConfig, currentConfigErrors
+}
 
-// New creates a new config object
-func New() (*config, error) {
-	conf := defaultConfig
+// Load loads the configuration from the file if a preferred config is not provided
+//
+// This is useful when Munch is being used as a package and the user wants to dynamically pass their own configuration
+func Load(c *config) (*config, error) {
+	currentConfig = NewDefaultConfig() // Assign it to config.currentConfig
 
-	pwd, err := os.Getwd()
-	cobra.CheckErr(err)
-	viper.AddConfigPath(pwd)
-	viper.AddConfigPath(".")
-	viper.SetConfigType("yml")
-	viper.SetConfigName("munch")
+	// Use the provided config if it's not nil
+	if c != nil {
+		currentConfig = c
+	}
 
-	if err := viper.ReadInConfig(); err != nil {
-		slog.Warn("Config file not found, using default config", "error", err)
-		return defaultConfig, nil 
+	// Load the config from the file if it hasn't been provided
+	if c == nil {
+		pwd, err := os.Getwd()
+		cobra.CheckErr(err)
+		viper.AddConfigPath(pwd)
+		viper.AddConfigPath(".")
+		viper.SetConfigType("yml")
+		viper.SetConfigName("munch")
+
+		if err := viper.ReadInConfig(); err != nil {
+			slog.Warn("Config file not found, using default config", "error", err)
+			return currentConfig, nil
+		}
+
+		if err := viper.Unmarshal(currentConfig); err != nil {
+			slog.Error(e.FAIL, "err", err, "description", "Couldn't parse config, using default config")
+			return currentConfig, nil
+		}
 	}
 
-	if err := viper.Unmarshal(conf); err != nil {
-		slog.Error(e.FAIL, "err", err, "description", "Couldn't parse config, using default config")
-		return defaultConfig, nil 
+	// Check for critical values, but do not panic. Instead, return an error
+	if err := validateCriticalFields(currentConfig, false); err != nil {
+		currentConfigErrors = err
+		return currentConfig, currentConfigErrors
 	}
-	// Check for critical values and panic if missing
-	validateCriticalFields(conf)
 
-	return conf, nil
+	currentConfigErrors = nil // Reset the errors
+	return currentConfig, currentConfigErrors
+}
+
+// CriticalError represents a custom error type for critical config errors
+type CriticalError struct {
+	Field   string
+	Message string
+}
+
+// Error implements the error interface for CriticalError
+func (e *CriticalError) Error() string {
+	return "Critical config value missing: " + e.Field + " - " + e.Message
+}
+
+// CriticalErrors represents a collection of critical validation errors
+type CriticalErrors struct {
+	Errors []error
+}
+
+// Error implements the error interface for CriticalValidationErrors
+func (e *CriticalErrors) Error() string {
+	var errMsg string
+	for _, err := range e.Errors {
+		errMsg += err.Error()
+	}
+	return errMsg
 }
+
 // validateCriticalFields checks for critical parameters in the configuration
-func validateCriticalFields(conf *config) {
+func validateCriticalFields(conf *config, isPanic bool) error {
```

Comment:
When will we decide to set isPanic to be true? Could you give an example?

Reply:
Perhaps when Munch is kick started as a http/grpc server, to ensure reliable responses, the objective could be to fail hard whenever config loads aren't validated; rather than expecting the calling functions to panic. That may be too many places that will have to handle similar checks, so it is better done here. 

This is necessary when in-place config changes are rolled out for a web server on production that it fails right away rather than the delayed panic. Waiting for an inflight request of some route to trigger the panic is an ineffective way to mark a deployment as 'success'.

Reply:
I see